confluent_kafka/schema_registry/avro.py

```diff
@@ -0,0 +1,329 @@
+        :py:func:`SerializingProducer.produce()` or by registering a `to_dict`
+        callable with the AvroSerializer.
+
+        See ``avroproducer.py`` in the examples directory for more details on
+        usage.
+
+    Args:
+        schema_registry_client (SchemaRegistryClient): Schema Registry client.
+
+        schema_str (str): Avro Schema declaration.
+
+        to_dict (callable): Callable(object) -> dict. Converts object to a dict.
+
+        conf (dict): AvroSerializer configuration.
+
+    .. _Subject name strategy:
+        https://docs.confluent.io/current/schema-registry/serializer-formatter.html#subject-name-strategy
+
+    .. _Schema declaration:
+        https://avro.apache.org/docs/current/spec.html#schemas
+
+    """  # noqa: E501
+    __slots__ = ['_hash', 'auto_register', 'known_subjects', 'parsed_schema',
+                 'registry', 'schema', 'schema_id', 'schema_name',
+                 'subject_name_func', 'to_dict']
+
+    # default configuration
+    _default = {'auto.register.schemas': True,
+                'subject.name.strategy': topic_subject_name_strategy}
+
+    def __init__(self, schema_registry_client, schema_str,
+                 to_dict=None, conf=None):
+        self.registry = schema_registry_client
+        self.schema_id = None
+        # Avoid calling registry if schema is known to be registered
+        self.known_subjects = set()
+
+        if to_dict is not None and not callable(to_dict):
+            raise ValueError("to_dict must  be callable  with the signature"
+                             " to_dict(object, Serialization Context)->dict")
+
+        self.to_dict = to_dict
+
+        # handle configuration
+        conf_copy = self._default.copy()
+        if conf is not None:
+            conf_copy.update(conf)
+
+        self.auto_register = conf_copy.pop('auto.register.schemas')
+
+        if not isinstance(self.auto_register, bool):
+            raise ValueError("auto.register.schemas must be a boolean value")
+
+        subject_name_func = conf_copy.pop('subject.name.strategy')
+
+        if not callable(subject_name_func):
+            raise ValueError("subject.name.strategy must be callable.")
+
+        self.subject_name_func = subject_name_func
+
+        if len(conf_copy) > 0:
+            raise ValueError("Unrecognized property(ies) {}"
+                             .format(conf_copy.keys()))
+
+        # convert schema_str to Schema instance
+        schema = _schema_loads(schema_str)
+        schema_dict = loads(schema.schema_str)
+        parsed_schema = parse_schema(schema_dict)
+        #  https://github.com/fastavro/fastavro/issues/415
+        schema_name = parsed_schema.get('name', schema_dict['type'])
+
+        self.schema = schema
+        self.schema_name = schema_name
+        self.parsed_schema = parsed_schema
+        self._hash = hash(schema.schema_str)
+
+    def __hash__(self):
+        return self._hash
```

Comment:
why does `__hash__` need to be defined for `AvroSerializer`? also, if it did, it probably should incorporate other member variables as well.